### DISCUSSION TITLE
Introducing a New RabbitMQ Toolkit for SuperAGI

### DISCUSSION THREAD

**Discussion by zhound420** [TransformerOptimus/SuperAGI#610] [Ideas]
Hello SuperAGI community,

First let me be clear that I'm not a developer by trade. I did not find much documentation here but  with the help of gpt4  I was able to write a new toolkit for the SuperAGI project - a RabbitMQ Toolkit. **This toolkit is designed to facilitate communication and collaboration between agents using RabbitMQ**, a robust and widely-used open-source message broker.

The RabbitMQ Toolkit includes operations for sending and receiving messages to and from RabbitMQ queues. It's designed to be easy to use, with a simple API that integrates seamlessly with the rest of the SuperAGI project. The toolkit also includes robust error handling and connection retries to ensure reliable operation.

Here's a brief overview of the toolkit's functionality:

- Sending messages to a RabbitMQ queue
- Receiving messages from a RabbitMQ queue
- Robust error handling and connection retries
- Asynchronous operations for improved performance

The toolkit is ready for testing and I am eager to integrate it into the SuperAGI project. However, I am currently facing a challenge in registering the toolkit as a valid toolkit in the application. Specifically, I am having trouble with the 'organisation', 'session', and 'tool_name_to_toolkit' variables in the init_tools and init_toolkits functions in tool_helper.py.

I would greatly appreciate any guidance or assistance the community can provide in resolving this issue. I believe that the RabbitMQ Toolkit will be a valuable addition to the SuperAGI project, and I am excited to see it integrated into the system.

Thank you in advance for your help and I look forward to your feedback.

**Comment by zhound420** [comment]
I have strictly followed the method from existing toolkits. As I see the toolkits are recognized from the "tools" directory and the "tool_helper.py" scans and registers the toolkits. 

I have it running with no errors, logs say that all tools are registered. But my new tool does not show up in the gui anywhere. Seems like maybe the toolkit is not getting registered but logs don't say otherwise.

**Reply by luciferlinx101** [reply]
Hey once you make a new tool and put it inside the superagi/tools folder it should be visible if you follow the right pattern on the next build using ```docker compose up --build```. If you still face the same issue please raise a pull request would love to help out.

**Reply by zhound420** [reply]
Thank you @luciferlinx101. I also joined the discord. Here are the logs: 

```
superagi-backend-1          | 2023-07-03 16:11:13 UTC - Super AGI - INFO - [/app/main.py:136] - Running Startup tasks
superagi-backend-1          | 2023-07-03 16:11:13 UTC - Super AGI - INFO - [/app/main.py:140] - User(id=1, name='SuperAGI User', email='super6@agi.com', password='pass@123',organisation_id=1)
superagi-backend-1          | 2023-07-03 16:11:13 UTC - Super AGI - INFO - [/app/main.py:143] - Organisation(id=1, name='Default Organization - 1')
superagi-backend-1          | 2023-07-03 16:11:15 UTC - Super AGI - INFO - [/app/superagi/helper/tool_helper.py:249] - Toolkits Registered Successfully for Organisation ID : 1!
superagi-backend-1          | 2023-07-03 16:11:16 UTC - Super AGI - INFO - [/app/superagi/helper/tool_helper.py:249] - Toolkits Registered Successfully for Organisation ID : 1!
superagi-backend-1          | 2023-07-03 16:11:16 UTC - Super AGI - INFO - [/app/main.py:257] - Successfully registered local toolkits for all Organisations!
superagi-backend-1          | INFO:     Application startup complete.
```

All other tools are showing up fine in the UI except for the new rabbitmq toolkit. The structure of the toolkit scripts follow exactly the same method as other toolkits. Kind of hitting a wall here.  Thanks for the help.

**Reply by luciferlinx101** [reply]
Hey can you raise a pull request for the same would love to figure out what's going wrong.

**Reply by zhound420** [reply]
Thank you @luciferlinx101 , sorry I'm fairly new to github collaboration. Just created a pull request, please bear with me if I make any mistakes. Really appreciate your help.

**Reply by luciferlinx101** [reply]
Sure we will review and let you know.

**Reply by zhound420** [reply]
@luciferlinx101 . So sorry, I was trying to update my fork and ended up deleting it which appears to also has removed my pull request. Anyway now that we have external tools method in v.07 hopefully you can still help me with it as an external toolkit. I did put the updates on discord, we can continue there for support, really appreciate your help.

**Reply by luciferlinx101** [reply]
Hey sure i have replied there.

**Reply by luciferlinx101** [reply]
@zhound420 Any help needed further from my end?

**Reply by zhound420** [reply]
@luciferlinx101 Thank you. I am working on it in my free time so taking me a bit of time. I have the toolkit registering properly and talking to the rabbitmq server. Working on:

1. ENV variable for SERVER_NAME doesn't seem to be passing to the container
2. Improving the _execute methods to incorporate better agent utilization
3. Split the different tool actions into separate files. 

Thanks again. I'll reach out if I get stuck.
